File: ios/Stopwatch/domain/use_cases/impl/NewLapUseCaseImpl.swift
class NewLapUseCaseImpl<MSS>: NewLapUseCase
where MSS: MutableStateStore, MSS.State == StopwatchState
{
    private let store: MSS
    private let calculateLapsStatuses: CalculateLapsStatuses
    
    init(_ store: MSS, _ calculateLapsStatuses: CalculateLapsStatuses) {
        self.store = store
        self.calculateLapsStatuses = calculateLapsStatuses
    }
    
    func execute() async {
        await store.update {currentState in
            var laps = currentState.completedLaps.map { $0 }
            
            laps.append(
                Lap(
                    index: currentState.completedLaps.count + 1,
                    milliseconds: currentState.milliseconds - currentState.completedLapsMilliseconds,
                    status: .done
                )
            )

            return StopwatchState(
                status: currentState.status,
                milliseconds: currentState.milliseconds,
                completedLaps: calculateLapsStatuses.execute(laps),
                completedLapsMilliseconds: currentState.milliseconds
            )
        }
    }
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/domain/use_cases/impl/NewLapUseCaseImpl.kt
package com.rafaeltmbr.stopwatch.domain.use_cases.impl

import com.rafaeltmbr.stopwatch.domain.data.stores.MutableStateStore
import com.rafaeltmbr.stopwatch.domain.entities.Lap
import com.rafaeltmbr.stopwatch.domain.entities.StopwatchState
import com.rafaeltmbr.stopwatch.domain.use_cases.NewLapUseCase
import com.rafaeltmbr.stopwatch.domain.utils.CalculateLapsStatuses

class NewLapUseCaseImpl(
    private val store: MutableStateStore<StopwatchState>,
    private val calculateLapsStatuses: CalculateLapsStatuses
) : NewLapUseCase {
    override suspend fun execute() {
        store.update {
            val newLap = Lap(
                index = it.laps.size + 1,
                milliseconds = 0L,
                status = Lap.Status.CURRENT
            )

            val mergedLaps = it.laps + newLap

            it.copy(laps = calculateLapsStatuses.execute(mergedLaps))
        }
    }
}